app/models/task.rb

```diff
@@ -4,6 +4,42 @@ class Task < ApplicationRecord
 
   validates :name, presence: true
 
+  after_create_commit :notify_create
+  after_create_commit :notify_almost_streak, if: :almost_streak?
+  after_update_commit :notify_almost_streak, if: :almost_streak?
+
+  before_destroy :clean_up_notifications
+
+  has_noticed_notifications model_name: 'Notification'
+
   scope :completed, -> { where(complete: true) }
   scope :uncompleted, -> { where(complete: false) }
+
+  private
+
+  ### methods for notice ###
+  #
+  # Checks if the habits is in "almost streak" state
+  # when true then notify_almost_streak
+  def almost_streak?
```

Comment:
Looks like a good candidate for `Streakable` concern to make the code more "DRY"